3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,217 @@
+   * Gets a handle to the window (or testing object that mocks window)
+   * @returns {!Window}
+   */
+  getWindow() {
+    return this.win_;
+  }
+
+  /**
+   * Sends a message from the third-party vendor's metrics-collection page back
+   * to the creative.
+   * @private
+   */
+  sendReadyMessageToCreative_() {
+    this.iframeMessagingClient_.sendMessage(AMP_ANALYTICS_3P_MESSAGE_TYPE.READY,
+      {});
+  }
+
+  /**
+   * Sends a message back to the host window
+   * @param {string} type The type of message to send.
+   * @param {(ampAnalytics3pReadyMessage|ampAnalytics3pResponse)} opt_payload
+   * The payload of message to send.
+   */
+  sendMessage(type, opt_payload) {
+    this.iframeMessagingClient_.sendMessage(type, opt_payload);
+  }
+}
+
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
+   * @param {!string} creativeId
+   */
+  constructor(parent, creativeId) {
+    /** @private {!AmpAnalytics3pMessageRouter} */
+    this.parent_ = parent;
+
+    /**
+     * @private {string}
+     */
+    this.creativeId_ = creativeId;
+
+    /**
+     * @private {string}
+     */
+    this.extraData_ = null;
+
+    /**
+     * @private {!Array<Object<string,*>>}
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * @private {!Array<!function(Array)>}
+     */
+    this.eventListeners_ = [];
+
+    if (this.parent_.getWindow().onNewAmpAnalyticsInstance) {
+      this.parent_.getWindow().onNewAmpAnalyticsInstance(this);
+    } else {
+      dev().error(TAG_, 'Vendor page must implement onNewAmpAnalyticsInstance' +
+        ' prior to loading library script.');
+    }
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur
+   * @param {!function(!Array)} listener A function that takes an array of event
+   * strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    this.eventListeners_.push(listener);
+  }
+
+  /**
+   * Receives a message from a creative for the cross-domain iframe
+   * @param {Object<string,*>} message The message that was received
+   */
+  receiveMessage(message) {
+    if (message[AMP_ANALYTICS_3P_MESSAGE_TYPE.NEW]) {
+      this.extraData_ = message[AMP_ANALYTICS_3P_MESSAGE_TYPE.NEW];
+    } else {
+      this.receivedMessagesQueue_.push(message);
+    }
+  }
+
+  /**
+   * If an event listener has been created on the third-party vendor's
+   * metrics-collection page, this method passes the queued events to that
+   * listener.
+   * If there is no event listener on the page yet, event messages will remain
+   * in the queue.
+   */
+  sendQueuedMessagesToListeners() {
+    if (this.receivedMessagesQueue_.length == 0 ||
+      this.eventListeners_.length == 0) {
+      return;
+    }
+    this.eventListeners_.forEach(listener => {
```

Comment:
this.eventListeners_.forEach(listener => listener(this. receivedMessagesQueue_));

We should consider wrapping each listener call with try/catch as otherwise the first throwing would not only cause the rest not to execute, it would cause the queue not to be cleared.